### PR TITLE
Add new membership inference attacks

### PR DESCRIPTION
**Pull request by abigailgold**
# Description

Added two membership inference attacks: one rule-based and one learned black-box attack. 
Added two Jupyter notebooks explaining how to apply both membership and attribute inference attacks.
Added dependency on pandas for preprocessing of nursery dataset used in new notebooks.

## Type of change

- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# Testing

Added tests for new membership attacks in tests/attacks/inference/test_membership_inference.py

**Test Configuration**:
- OS: MacOS
- Python version: 3.7
- ART version or commit number
- TensorFlow 2.1 / Keras 2.3.1 / PyTorch 0.1 

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -11,6 +11,7 @@ pydub==0.24.1
 resampy==0.2.2
 ffmpeg-python==0.2.0
 cma==2.7.0
+pandas==0.24.2
```
Comment (beat-buesser) [inline requirements.txt:14]: Is there a reason for using pandas 0.24 instead of pandas 1.x?

File: art/estimators/estimator.py [hunk 2]
```diff
@@ -512,6 +512,20 @@ def set_learning_phase(self, train: bool) -> None:
         """
         raise NotImplementedError
 
+    @abstractmethod
+    def loss(self, x: np.ndarray, y: np.ndarray, **kwargs) -> np.ndarray:
+        """
+        Compute the loss of the neural network for samples `x`.
```
Comment (beat-buesser) [inline art/estimators/estimator.py:518]: Estimators don't have to be neural networks.
```suggestion
        Compute the loss of the model for samples `x`.
```

File: art/estimators/classification/tensorflow.py [hunk 3]
```diff
@@ -907,6 +920,43 @@ def class_gradient(self, x: np.ndarray, label: Union[int, List[int], None] = Non
...
+                "defined."
+            )
+        prev_reduction = self._loss_object.reduction
+        self._loss_object.reduction = tf.keras.losses.Reduction.NONE
+
+        # Apply preprocessing
+        x_preprocessed, _ = self._apply_preprocessing(x, y, fit=False)
+
+        if tf.executing_eagerly():
+            # with tf.GradientTape() as tape:
```
Comment (beat-buesser) [inline art/estimators/classification/tensorflow.py:946]:
```suggestion
```

File: art/estimators/classification/tensorflow.py [hunk 4]
```diff
@@ -907,6 +920,43 @@ def class_gradient(self, x: np.ndarray, label: Union[int, List[int], None] = Non
...
+            x_preprocessed_tf = tf.convert_to_tensor(x_preprocessed)
+            # tape.watch(x_preprocessed_tf)
```
Comment (beat-buesser) [inline art/estimators/classification/tensorflow.py:948]:
```suggestion
```

File: art/attacks/inference/membership_inference.py [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+
+import logging
+from typing import Optional
+
+import numpy as np
+import torch
+import torch.nn as nn
+import torch.optim as optim
+from torch.utils.data.dataset import Dataset
+from torch.utils.data import DataLoader
```
Comment (beat-buesser) [inline art/attacks/inference/membership_inference.py:34]: To make ART framework independent, these imports have to be moved inside of the methods where they are required.

File: art/attacks/inference/membership_inference.py [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+
+from art.attacks import InferenceAttack
+from art.estimators.estimator import BaseEstimator, NeuralNetworkMixin
+from art.estimators.classification.classifier import ClassifierMixin, Classifier
+from art.utils import check_and_transform_label_format
+from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
```
Comment (beat-buesser) [inline art/attacks/inference/membership_inference.py:40]: For PEP8, please move the import of `sklearn` after the import of `numpy`.

File: art/attacks/inference/membership_inference.py [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+use_cuda = torch.cuda.is_available()
+
+
+def to_cuda(x):
+    if use_cuda:
+        x = x.cuda()
+    return x
+
+
+class AttackDataset(Dataset):
```
Comment (beat-buesser) [inline art/attacks/inference/membership_inference.py:53]: Could you please add a short docstring for this class?

File: art/attacks/inference/membership_inference.py [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+        return len(self.x1)
+
+    def __getitem__(self, idx):
+        if idx >= len(self.x1):
+            raise IndexError("Invalid Index")
+
+        return self.x1[idx], self.x2[idx], self.y[idx]
+
+
+class MembershipInferenceAttackModel(nn.Module):
```
Comment (beat-buesser) [inline art/attacks/inference/membership_inference.py:73]: Could you please add a short docstring for this class?

File: art/attacks/inference/membership_inference.py [hunk 9]
```diff
@@ -0,0 +1,346 @@
...
+
+
+class MembershipInferenceBlackBoxRuleBased(InferenceAttack):
+    """
+        Implementation of a simple, rule-based black-box membership inference attack.
+
+        This implementation uses the simple rule: if the model's prediction for a sample is correct, then it is a
+        member. Otherwise, it is not a member.
+    """
+    _estimator_requirements = [BaseEstimator]
```
Comment (beat-buesser) [inline art/attacks/inference/membership_inference.py:121]: Does `MembershipInferenceBlackBoxRuleBased` require classification models? If yes, we should add `ClassifierMixin` to `_estimator_requirements`.

File: art/attacks/inference/membership_inference.py [hunk 10]
```diff
@@ -0,0 +1,346 @@
...
+
+
+class MembershipInferenceBlackBox(InferenceAttack):
+    """
+        Implementation of a learned black-box membership inference attack.
+
+        This implementation can use as input to the learning process probabilities/logits or losses,
+        depending on the type of model and provided configuration.
+    """
+    _estimator_requirements = [BaseEstimator]
```
Comment (beat-buesser) [inline art/attacks/inference/membership_inference.py:160]: Does MembershipInferenceBlackBoxRuleBased require classification models? If yes, we should add ClassifierMixin to _estimator_requirements.
```suggestion
    _estimator_requirements = [BaseEstimator, ClassifierMixin]
```

File: art/attacks/inference/membership_inference.py [hunk 11]
```diff
@@ -0,0 +1,346 @@
...
+        return [1 if p == y[index] else 0 for index, p in enumerate(predictions)]
+
+
+class MembershipInferenceBlackBox(InferenceAttack):
+    """
+        Implementation of a learned black-box membership inference attack.
+
+        This implementation can use as input to the learning process probabilities/logits or losses,
+        depending on the type of model and provided configuration.
+    """
```
Comment (beat-buesser) [inline art/attacks/inference/membership_inference.py:159]:
```suggestion
    """
    
    attack_params = InferenceAttack.attack_params + [
        "input_type",
        "attack_model_type",
        "attack_model",
    ]
```

File: art/attacks/inference/membership_inference.py [hunk 12]
```diff
@@ -0,0 +1,346 @@
...
+                    self.attack_model = MembershipInferenceAttackModel(classifier.nb_classes, 1)
+                self.epochs = 100
+                self.bs = 100
+                self.lr = 0.0001
+            elif attack_model_type == 'rf':
+                self.attack_model = RandomForestClassifier()
+            elif attack_model_type == 'gb':
+                self.attack_model = GradientBoostingClassifier()
+            else:
+                raise ValueError("Illegal value for parameter `attack_model_type`.")
```
Comment (beat-buesser) [inline art/attacks/inference/membership_inference.py:205]: After setting all attributes we should run method `_check_params` which contains type and value checks for each attribute if possible. A method `_check_params` needs to be implemented by this class.
```suggestion
                raise ValueError("Illegal value for parameter `attack_model_type`.")
        self._check_params()
```

File: art/attacks/inference/membership_inference.py [hunk 13]
```diff
@@ -0,0 +1,346 @@
...
+        Infer sensitive properties (attributes, membership training records) from the targeted estimator. This method
+        should be overridden by all concrete inference attack implementations.
+
+        :param x: An array with reference inputs to be used in the attack.
+        :param y: Labels for `x`. This parameter is only used by some of the attacks.
+        :return: An array holding the inferred properties.
+        """
+        raise NotImplementedError
+        # if NeuralNetworkMixin in type(self.estimator).__mro__:
+        #     activations = self.estimator.get_activations(x, self.estimator.layer_names[-1], batch_size=self.bs)
```
Comment (beat-buesser) [inline art/attacks/inference/membership_inference.py:346]: How is this class used?

File: art/estimators/classification/keras.py [hunk 14]
```diff
@@ -237,7 +242,7 @@ def _initialize_params(
             "__name__" in dir(loss_function)
             and loss_function.__name__
             in ["categorical_hinge", "categorical_crossentropy", "binary_crossentropy", "kullback_leibler_divergence",]
-        ) or (self.is_tensorflow and flag_is_instance):
```
Comment (beat-buesser) [inline art/estimators/classification/keras.py:240]: Is removing `self.is_tensorflow` required for this pull request? I'm still concerned that it was required here.

File: art/estimators/generation/tensorflow.py [hunk 15]
```diff
@@ -100,8 +100,18 @@ def __init__(
             # TODO do the same thing for all not None variables
         self._sess = sess
 
-    def loss(self, z, image_adv):
-        pass
```
Comment (beat-buesser) [inline art/estimators/generation/tensorflow.py:104]: @killianlevacher Did you need this method `loss(self, z, image_adv)` for something? Can we safely replace it with the new method below?

File: tests/attacks/inference/test_membership_inference.py [hunk 16]
```diff
@@ -0,0 +1,363 @@
...
+
+
+logger = logging.getLogger(__name__)
+attack_train_ratio = 0.5
+num_classes_iris = 3
+num_classes_mnist = 10
+
+
+@pytest.fixture()
+def fix_get_mnist_subset(get_mnist_dataset):
```
Comment (killianlevacher) [inline tests/attacks/inference/test_membership_inference.py:47]: this fixture needs to be removed and instead the project-wide `get_default_mnist_subset` should be used

File: tests/attacks/inference/test_membership_inference.py [hunk 17]
```diff
@@ -0,0 +1,363 @@
...
+        logging.warning("Couldn't perform this test because no classifier is defined")
+        return
+
+    (x_train, y_train), (x_test, y_test) = get_iris_dataset
+
+    for classifier in classifier_list:
+        # print(type(classifier).__name__)
+        attack = MembershipInferenceBlackBoxRuleBased(classifier)
+        # infer attacked feature
+        inferred_train = attack.infer(x_train, y_train)
```
Comment (killianlevacher) [inline tests/attacks/inference/test_membership_inference.py:90]: lines from 90 to 97 are pratically identical across all tests. For maintenance purposes, it would be better to use a `_backend_test_something` method was used. An example of this pattern can be found at `tests.attacks.evasion.test_fast_gradient`

File: tests/attacks/inference/test_membership_inference.py [hunk 18]
```diff
@@ -0,0 +1,363 @@
...
+    classifier_list = get_tabular_classifier_list(MembershipInferenceBlackBox)
+    if not classifier_list:
+        logging.warning("Couldn't perform this test because no classifier is defined")
+        return
+
+    (x_train, y_train), (x_test, y_test) = get_iris_dataset
+    attack_train_size = int(len(x_train) * attack_train_ratio)
+    attack_test_size = int(len(x_test) * attack_train_ratio)
+
+    model_types = ['nn', 'rf', 'gb']
```
Comment (killianlevacher) [inline tests/attacks/inference/test_membership_inference.py:136]: the `model_types` parameters should be set as a pytest marker parameter. See `@pytest.mark.parametrize("norm", [np.inf, 1, 2])` in test `tests.attacks.evasion.test_fast-gradient.test_norm_images`

File: tests/attacks/inference/test_membership_inference.py [hunk 19]
```diff
@@ -0,0 +1,363 @@
...
```
Comment (killianlevacher) [inline tests/attacks/inference/test_membership_inference.py:165]: same thing as above

File: requirements.txt (hunk 1 above)
Comment (abigailgold) [reply requirements.txt:14]: That's just the version I already had installed. I can test with version 1.1 and see if it works as well.

File: art/attacks/inference/membership_inference.py (hunk 13 above)
Comment (abigailgold) [reply art/attacks/inference/membership_inference.py:346]: It is not at the moment. It's just a placeholder, I didn't have time to implement it for this release. I guess I should remove it then, right?

File: art/estimators/classification/keras.py (hunk 14 above)
Comment (abigailgold) [reply art/estimators/classification/keras.py:240]: Yes, it is required for the tests to pass. Look at the following elif (line 250), it does a very similar check (checks if loss_function is an instance of a particular class) but does not also force self.is_tensorflow to be true. When using "regular" keras (i.e., import keras and not import tensorflow.keras) these loss classes still exist and when using them without this change you get to the raise ValueError("Loss function not recognised.").

File: tests/attacks/inference/test_membership_inference.py [hunk 20]
```diff
@@ -0,0 +1,363 @@
...
+
+
+@pytest.mark.only_with_platform("keras")
+@pytest.mark.skipif(keras.__version__.startswith('2.2'), reason="requires Keras 2.3.0 or higher")
+def test_black_box_keras_loss(get_iris_dataset):
+    (x_train, y_train), (x_test, y_test) = get_iris_dataset
+    attack_train_size = int(len(x_train) * attack_train_ratio)
+    attack_test_size = int(len(x_test) * attack_train_ratio)
+
+    model = keras.models.Sequential()
```
[the remark this answers is not in the document]
Comment (abigailgold) [reply tests/attacks/inference/test_membership_inference.py:193]: So there is actually nothing special about the model itself, but I needed to be able to test with both a string-defined loss (e.g., loss='categorical_crossentropy') and a class-based loss (e.g., loss=keras.losses.CategoricalCrossentropy()). The  get_tabular_classifier_list fixture only provides the first option. So how can I add an additional keras tabular classifier that uses a class-based loss object?

File: tests/attacks/inference/test_membership_inference.py (hunk 17 above)
Comment (abigailgold) [reply tests/attacks/inference/test_membership_inference.py:90]: So are you saying I should put the _backend_test_something method in attacks/utils.py? Even if it's only used in this test file? Or should I add it within test_membership_inference.py?

File: tests/attacks/inference/test_membership_inference.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (killianlevacher) [reply tests/attacks/inference/test_membership_inference.py:193]: a ok that makes sense yes I understand the reason now. @beat-buesser I guess the only way we could make this test framework independent is if we enabled `get_tabular_classifier_list` to get a loss input? That might be something to look towards in the future. At a minimum, I would suggest then maybe to put a comment explaining the reason why these local models were needed (to help out future refactoring). Also if these these keras models will be left, there would be a need to add a `skip_by_platform` marker to the test to make sure it isn't ran by other frameworks

File: tests/attacks/inference/test_membership_inference.py (hunk 17 above)
Comment (killianlevacher) [reply tests/attacks/inference/test_membership_inference.py:90]: yes I think leaving it in `test_membership_inference.py` is probably a better idea since it's only used here

File: art/attacks/inference/membership_inference.py (hunk 13 above)
Comment (beat-buesser) [reply art/attacks/inference/membership_inference.py:346]: Yes, removing was my first thought because it does not implement methods. But if you would like to claim the name space and have plans to implement it in the very near future you can leave it, in that case I would ask to add a comment describing the plans.

**Comment by abigailgold** [comment]
Opened a new pull request from a clean branch (573)
